packages/@romejs/js-compiler/lint/rules/react/noDirectMutationState.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {Path} from '@romejs/js-compiler';
+import {AnyNode} from '@romejs/js-ast';
+import {descriptions} from '@romejs/diagnostics';
+import {doesNodeMatchPattern} from '@romejs/js-ast-utils';
+
+export default {
+  name: 'noDirectMutationState',
+  enter(path: Path): AnyNode {
+    const {node} = path;
+
+    if (
+      node.type === 'ClassDeclaration' &&
+      node.meta.type === 'ClassHead' &&
+      node.meta.superClass?.type === 'MemberExpression' &&
+      doesNodeMatchPattern(node.meta.superClass, 'React.Component')
+    ) {
+      for (let bodyNode of node.meta.body) {
+        if (
+          (bodyNode.type === 'ClassMethod' || bodyNode.type === 'ClassPrivateMethod') &&
+          bodyNode.kind !== 'constructor'
+        ) {
+          for (let bodyBodyNode of bodyNode.body.body) {
+            if (
+              bodyBodyNode.type === 'ExpressionStatement' &&
+              bodyBodyNode.expression.type === 'AssignmentExpression' &&
+              bodyBodyNode.expression.left.type === 'MemberExpression' &&
+              bodyBodyNode.expression.left.object.type === 'MemberExpression' &&
+              bodyBodyNode.expression.left.object.object.type === 'ThisExpression' &&
+              bodyBodyNode.expression.left.object.property.value.type === 'Identifier' &&
+              bodyBodyNode.expression.left.object.property.value.name === 'state'
```

Comment:
With #396 this can now just be:

```suggestion
              doesNodeMatchPattern(bodyBodyNode.expression.left, 'this.state')
```

Or if this check is valuable you could also do:

```
doesNodeMatchPattern(bodyBodyNode.expression.left, 'this.state.**')
```

Reply:
It seems like doesNodeMatchPattern is having difficulty with `ThisExpressions`.  I've tried a bit and can't get it to work, but I hope to figure out why.